File: ocaml-interpreter/test/lexer_tests.ml
open OUnit2
open Interpreter.Lexer
open Interpreter.Token

let lexer_tests () =
  let test_next_token _ =
    let input = 
      "let five = 5;"
      ^ "let ten = 10;"
      ^ "let add = fn(x, y) {"
      ^ "x + y;"
      ^ "};"
      ^ "let result = add(five, ten);"
      ^ "!-/*5;"
      ^ "5 < 10 > 5;"
      ^ "if (5 < 10) {"
      ^ "return true;"
      ^ "} else {"
      ^ "return false;"
      ^ "}"
      ^ "10 == 10;"
      ^ "10 != 9;"
    in

    let tests = [
      Let; Ident "five"; Assign; Int 5; Semicolon;
      Let; Ident "ten"; Assign; Int 10; Semicolon;
      Let; Ident "add"; Assign; Function; LeftParen; Ident "x"; Comma; Ident "y"; RightParen; LeftCurly;
      Ident "x"; Plus; Ident "y"; Semicolon;
      RightCurly; Semicolon;
      Let;Ident "result";
      Assign; Ident "add"; LeftParen; Ident "five"; Comma; Ident "ten"; RightParen; Semicolon;
      Bang; Minus; Slash; Asterisk; Int 5; Semicolon;
      Int 5; LessThan; Int 10; GreaterThan; Int 5; Semicolon;
      If; LeftParen; Int 5; LessThan; Int 10; RightParen; LeftCurly;
      Return; True; Semicolon;
      RightCurly; Else; LeftCurly;
      Return; False; Semicolon;
      RightCurly;
      Int 10; Equal; Int 10; Semicolon;
      Int 10; NotEqual; Int 9; Semicolon;
      Eof;
    ] in

    let rec run_tests lexer tests =
      let token, lexer = next_token lexer in
      match tests with
        | [] -> ()
        | tok :: tests -> 
          assert_equal tok token;
          run_tests lexer tests
    in
    run_tests (make input) tests
  in
  
  "Lexer Tests" >::: [
    "next_token" >:: test_next_token;
  ]

let () =
  run_test_tt_main (lexer_tests ())
   

File: swift-interpreter/Tests/InterpreterTests/LexerTests.swift
import XCTest
@testable import Interpreter

final class LexerTests: XCTestCase {
    func testNextToken() {
				let input = """
					let five = 5;
					let ten = 10;

					let add = fn(x, y) {
					x + y;
					};

					let result = add(five, ten);
				"""
				
        var lexer = Lexer.init(input);
				let tests: [Token] = [
					.let,
					.ident("five"),
					.assign,
					.int(5),
					.semicolon,
					.let,
					.ident("ten"),
					.assign,
					.int(10),
					.semicolon,
					.let,
					.ident("add"),
					.assign,
					.function,
					.lparen,
					.ident("x"),
					.comma,
					.ident("y"),
					.rparen,
					.lbrace,
					.ident("x"),
					.plus,
					.ident("y"),
					.semicolon,
					.rbrace,
					.semicolon,
					.let,
					.ident("result"),
					.assign,
					.ident("add"),
					.lparen,
					.ident("five"),
					.comma,
					.ident("ten"),
					.rparen,
					.semicolon,
					.eof,
				]
				for test in tests {
					XCTAssertEqual(test, lexer.nextToken())
				}
    }

    static var allTests = [
        ("testNextToken", testNextToken),
    ]
}
